### PR TITLE
refine unix domain socket

### PR DESCRIPTION
**Pull request by lucifer9**
This PR moves Unix Domain Socket to the place it should be. UDS is no longer a transport.

Some limitations:
- Only Unix Domain Socket of type SOCK_STREAM is supported, so only tcp, http, websocket transports (w/ or w/o tls,xtls) can use UDS now. UDP based transports are not supported.
- Only Inbound support(only listener, no dialer). It seems weird to use UDS in Outbound.
- This PR enables vmess, VLESS, trojan to work with UDS. Other proxy protocols seem unnecessary to use UDS, but it is easy to add.

**It is compatible with previous configuration.** But it is recommended to gradually start experimenting with the new configuration, especially when you want to use DS together with h2/ws transport.

Configuration changes:

- Use non-ip values begins with "@" or "/" in "Listen" to listen on UDS, "port" will be ingored then.
``` 
"Listen": "/tmp/test.sock" 
```
- UDS begins with "@" means abstract UDS in Linux, "@@" means padding(to work with haproxy), thanks to @rprx . On other platforms it is still normal UDS.
``` 
"Listen": "@test.sock" // abstract unix domain, linux only
"Listen": "@@test.sock" // adbstract unix domain with padding, to work with haproxy, linux only
```

- Add "acceptProxyProtocol" to "sockopt". It is implemented only once for all possible transports in SystemListener now and is more generic.
```
"sockopt": {
      "acceptProxyProtocol": true
}
```

### CODE REVIEW

**Comment by kslr** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- app/proxyman/inbound/always.go  3
- common/net/network.go  1
- transport/internet/filelocker_other.go  4
- transport/internet/tcp/hub.go  4
- transport/internet/filelocker_windows.go  1
- transport/internet/websocket/hub.go  4
- transport/internet/system_listener.go  5
- transport/internet/http/hub.go  3
- common/net/destination.go  1
         

Clones added
============
- app/proxyman/inbound/always.go  2
- transport/internet/filelocker_other.go  2
- transport/internet/tcp/hub.go  1
- transport/internet/websocket/hub.go  2
- app/proxyman/inbound/worker.go  6
- transport/internet/http/hub.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/v2fly/v2ray-core/pullRequest?prid=6419837&bid=20798317)

File: transport/internet/filelocker_windows.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package internet
+
+func (fl *FileLocker) Acquire() error {
```
Comment (kslr) [inline transport/internet/filelocker_windows.go:3]: Codacy found an issue: [exported method FileLocker.Acquire should have comment or be unexported](https://app.codacy.com/gh/v2fly/v2ray-core/pullRequest?prid=6419837)

File: transport/internet/filelocker_other.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	}
+	if err := unix.Flock(int(f.Fd()), unix.LOCK_EX); err != nil {
+		f.Close()
+		return newError("failed to lock file: ", fl.path).Base(err)
+	}
+	fl.file = f
+	return nil
+}
+
+func (fl *FileLocker) Release() {
```
Comment (kslr) [inline transport/internet/filelocker_other.go:24]: Codacy found an issue: [exported method FileLocker.Release should have comment or be unexported](https://app.codacy.com/gh/v2fly/v2ray-core/pullRequest?prid=6419837)

File: transport/internet/filelocker.go [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package internet
+
+import (
+	"os"
+)
+
+type FileLocker struct {
```
Comment (kslr) [inline transport/internet/filelocker.go:7]: Codacy found an issue: [exported type FileLocker should have comment or be unexported](https://app.codacy.com/gh/v2fly/v2ray-core/pullRequest?prid=6419837)

File: transport/internet/filelocker_windows.go [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+	return nil
+}
+
+func (fl *FileLocker) Release() {
```
Comment (kslr) [inline transport/internet/filelocker_windows.go:7]: Codacy found an issue: [exported method FileLocker.Release should have comment or be unexported](https://app.codacy.com/gh/v2fly/v2ray-core/pullRequest?prid=6419837)

File: transport/internet/filelocker_other.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+package internet
+
+import (
+	"os"
+
+	"golang.org/x/sys/unix"
+)
+
+func (fl *FileLocker) Acquire() error {
```
Comment (kslr) [inline transport/internet/filelocker_other.go:11]: Codacy found an issue: [exported method FileLocker.Acquire should have comment or be unexported](https://app.codacy.com/gh/v2fly/v2ray-core/pullRequest?prid=6419837)

File: transport/internet/tcp_hub.go [hunk 6]
```diff
@@ -27,6 +27,26 @@ type Listener interface {
 	Addr() net.Addr
 }
 
+func ListenUnix(ctx context.Context, address net.Address, settings *MemoryStreamConfig, handler ConnHandler) (Listener, error) {
```
Comment (kslr) [inline transport/internet/tcp_hub.go:30]: Codacy found an issue: [exported function ListenUnix should have comment or be unexported](https://app.codacy.com/gh/v2fly/v2ray-core/pullRequest?prid=6419837)

File: transport/internet/domainsocket/errgen.go [hunk 7]
```diff
@@ -1,3 +1,3 @@
 package domainsocket
 
-//go:generate go run v2ray.com/core/common/errors/errorgen
+//go:generate errorgen
```
Comment (Loyalsoldier) [inline transport/internet/domainsocket/errgen.go:3]: Please do NOT change this.

Comment (lucifer9) [reply transport/internet/domainsocket/errgen.go:3]: OK

**Comment by kslr** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- app/proxyman/inbound/always.go  3
- transport/internet/system_listener.go  5
- transport/internet/filelocker_other.go  4
- transport/internet/http/hub.go  3
- transport/internet/websocket/hub.go  4
- common/net/destination.go  1
- common/net/network.go  1
- transport/internet/tcp/hub.go  4
- transport/internet/filelocker_windows.go  1
         

Clones added
============
- app/proxyman/inbound/always.go  2
- transport/internet/filelocker_other.go  2
- transport/internet/http/hub.go  1
- transport/internet/websocket/hub.go  2
- transport/internet/tcp/hub.go  1
- app/proxyman/inbound/worker.go  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/v2fly/v2ray-core/pullRequest?prid=6419837&bid=20798317)

File: transport/internet/filelocker_other.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+	}
+	if err := unix.Flock(int(f.Fd()), unix.LOCK_EX); err != nil {
+		f.Close()
+		return newError("failed to lock file: ", fl.path).Base(err)
+	}
+	fl.file = f
+	return nil
+}
+
+// Realse lock
```
Comment (kslr) [inline transport/internet/filelocker_other.go:25]: Codacy found an issue: [comment on exported method FileLocker.Release should be of the form "Release ..."](https://app.codacy.com/gh/v2fly/v2ray-core/pullRequest?prid=6419837)

File: transport/internet/filelocker.go [hunk 9]
```diff
@@ -0,0 +1,11 @@
+package internet
+
+import (
+	"os"
+)
+
+// UDS access lock
```
Comment (kslr) [inline transport/internet/filelocker.go:7]: Codacy found an issue: [comment on exported type FileLocker should be of the form "FileLocker ..." (with optional leading article)](https://app.codacy.com/gh/v2fly/v2ray-core/pullRequest?prid=6419837)

**Comment by kslr** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- common/net/destination.go  1
- transport/internet/websocket/hub.go  4
- transport/internet/system_listener.go  5
- app/proxyman/inbound/always.go  3
- transport/internet/http/hub.go  3
- common/net/network.go  1
- transport/internet/filelocker_other.go  4
- transport/internet/tcp/hub.go  4
- transport/internet/filelocker_windows.go  1
         

Clones added
============
- transport/internet/websocket/hub.go  2
- app/proxyman/inbound/always.go  2
- transport/internet/http/hub.go  1
- app/proxyman/inbound/worker.go  6
- transport/internet/filelocker_other.go  2
- transport/internet/tcp/hub.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/v2fly/v2ray-core/pullRequest?prid=6419837&bid=20798317)

**Comment by rprx** [comment]
确定完全兼容现有的配置吗，优先级顺序是怎样的？

**Comment by lucifer9** [comment]
> 确定完全兼容现有的配置吗，优先级顺序是怎样的？

没有删改原本解析配置的部分，只有添加sockopt里面的acceptProxyProtocol，所以应该是完全兼容的。
优先级的话，原先的配置优先级最高。
- ds方面，因为原本的逻辑就是：streamSettings里面network是ds的话会无视之前的listen，port等配置。原有的ds直接调用go的ListenUnix，并没有用到system_listener，也就不会跟新的实现有交叉调用的风险。
- acceptProxyProtocol方面，在transport启动listener的时候，会同步自己的acceptProxyProtocol设置到sockopt，所以也是原先配置的优先级高。当然像是http这种本来就没有acceptProxyProtocol设置的tranport，就只有按照新的配置在sockopt里面启用Proxy Protocol，也就无所谓优先级了。

PS，稍后我补上log里面输出是否启用了Proxy Protocol

**Comment by kslr** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- transport/internet/tcp/hub.go  6
- transport/internet/filelocker_other.go  4
- app/proxyman/inbound/always.go  3
- common/net/destination.go  1
- transport/internet/system_listener.go  5
- transport/internet/websocket/hub.go  6
- transport/internet/filelocker_windows.go  1
- transport/internet/http/hub.go  5
- common/net/network.go  1
         

Clones added
============
- transport/internet/tcp/hub.go  2
- app/proxyman/inbound/worker.go  6
- transport/internet/filelocker_other.go  2
- app/proxyman/inbound/always.go  2
- transport/internet/websocket/hub.go  3
- transport/internet/http/hub.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/v2fly/v2ray-core/pullRequest?prid=6419837&bid=20798317)

**Review by rprx — APPROVED** [review]
还有其它补充吗

**Comment by lucifer9** [comment]
> 还有其它补充吗

没了。

**Comment by rprx** [comment]
关于 listen 的行为，建议改为只支持绝对路径，即至少需要以 `"/"` 开头，或者 `"@"`，其它的均按 TCP 处理。

即不特殊对待 localhost，因为可能有人自己定义了域名，这个行为和 fallbacks 也是一致的。

**Comment by rprx** [comment]
另外我不确定这个 PR 不写 listen 的行为是怎样的，但按上面的建议修改后，则一定可以和现在的行为保持一致。

**Comment by rprx** [comment]
可以顺便去掉 ds 的 acceptProxyProtocol，因为这个从来没有写进文档里，可能只有关注 PR 的开发者试过（个位数）。

**Comment by lucifer9** [comment]
> 关于 listen 的行为，建议改为只支持绝对路径，即至少需要以 `"/"` 开头，或者 `"@"`，其它的均按 TCP 处理。
> 
> 即不特殊对待 localhost，因为可能有人自己定义了域名，这个行为和 fallbacks 也是一致的。


> 另外我不确定这个 PR 不写 listen 的行为是怎样的，但按上面的建议修改后，则一定可以和现在的行为保持一致。



不写listen的话，原先是要求必须写port，然后去listen net.AnyIP。这一点跟PR里面是一样的。

**Comment by rprx** [comment]
@lucifer9 

好的，不过最好还是应该只支持 `"/"` 和 `"@"` 开头，否则 TCP，显示原本的报错信息（而且 ds 一般都写绝对路径）。

在 v2ray 各处支持 abstract 的情况下，这个更简短好用。

**Comment by kslr** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- transport/internet/filelocker_other.go  4
- transport/internet/system_listener.go  5
- transport/internet/http/hub.go  5
- transport/internet/websocket/hub.go  6
- transport/internet/filelocker_windows.go  1
- common/net/destination.go  1
- transport/internet/tcp/hub.go  6
- app/proxyman/inbound/always.go  3
- common/net/network.go  1
         

Complexity decreasing per file
==============================
+ transport/internet/domainsocket/listener.go  -1
         

Clones added
============
- transport/internet/domainsocket/listener.go  2
- transport/internet/filelocker_other.go  2
- transport/internet/http/hub.go  1
- transport/internet/websocket/hub.go  3
- transport/internet/tcp/hub.go  2
- app/proxyman/inbound/always.go  2
- app/proxyman/inbound/worker.go  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/v2fly/v2ray-core/pullRequest?prid=6419837&bid=20798317)

**Comment by lucifer9** [comment]
> @lucifer9
> 
> 好的，不过最好还是应该只支持 `"/"` 和 `"@"` 开头，否则 TCP，显示原本的报错信息（而且 ds 一般都写绝对路径）。
> 
> 在 v2ray 各处支持 abstract 的情况下，这个更简短好用。

OK，改过来了，顺便去掉了原来ds transport里面的acceptProxyProtocol

**Comment by rprx** [comment]
应该没什么问题了，测试 fallbacks dest 与 listen unix 对接 ok 吗？包括 abstract

**Comment by lucifer9** [comment]
> 应该没什么问题了，测试 fallbacks dest 与 listen unix 对接 ok 吗？包括 abstract

用终极配置的模版测试了没问题

**Comment by rprx** [comment]
@lucifer9 

是否有必要给 h2 加个 acceptProxyProtocol 以统一格式？

（另外之前你给 h2 加了识别 header，我还没写进文档里，它的优先级会比 PROXY protocol 更高）

**Comment by rprx** [comment]
或者干脆取消 h2 的 header 识别，因为它过不了 CDN，PROXY protocol 即可，但不知道可以反代 h2c 的软件是否支持。

**Comment by lucifer9** [comment]
acceptProxyProtocol的话，现在可以通过在sockopt里面设置，这样h2也就可以用了。以后如果准备去掉ds的transport的话，建议acceptProxyProtocol的设置也都统一到sockopt里面。因为以后可能还会有增加SOCK_DGRAM类型unix domain socket的需求，并且proxy protocol本身也可以用于udp和unix socket domain，所以如果以后有类似需求的话，现在就都放sockopt，会比单独每个transport加方便很多。
h2主要不是过cdn吧，比如我个人是用于前端跑haproxy做tls termination，后面接h2c的v2ray，这个时候header足够用了。

**Comment by rprx** [comment]
不是是否够用的问题，而是请求方可以自定义 header，如果它没被覆盖，恶意篡改就会成功。而 PROXY protocol 优先级比 header 低，也无法解决这个问题。

**Comment by lucifer9** [comment]
> 不是是否够用的问题，而是请求方可以自定义 header，如果它没被覆盖，恶意篡改就会成功。而 PROXY protocol 优先级比 header 低，也无法解决这个问题。

这样的话，h2跟ws都会有这种问题。当然代码里面可以判断一下做个优先级。
不过proxy protocol请求方也是可以自定义IP的啊

**Comment by rprx** [comment]
现在直连一般是 TCP，不存在这个问题，而 CDN 会给 WS 加 header，也解决了这个问题。

> 不过proxy protocol请求方也是可以自定义IP的啊

（当然只在内网接收 PROXY protocol）

**Comment by rprx** [comment]
我的想法是，因为现在 h2 支持接收 PROXY protocol 了，那么如果那些能反代 h2c 的软件也支持，则可以去掉 header 的识别。

**Comment by rprx** [comment]
（不过如果那些能反代 h2c 的软件都普遍设置加 header，则不存在这个问题）

**Comment by kslr** [comment]
怎么合并到master了

**Comment by lucifer9** [comment]
还是存在不少用户不通过cdn直接用nginx或者haproxy做tls termination来反代的，好像之前有说是效率更高，还是可以避免被识别啥的。
能反代 h2c 的软件里面，PROXY protocol的支持上，目前常用的，haproxy应该没问题，caddy2 看上去还不行https://github.com/caddyserver/caddy/issues/2724
另外，按照 https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For 的说法，X-Forwarded-For 默认行为是过一层代理加一个IP，真实client ip应该是第一个。目前代码里面也确实是取的forwardedAddrs[0]。不过这样也就没法避免伪造X-Forwarded-For。只要发起请求时候先随便写个X-Forwarded-For，不管过多少cdn，最后取得的就都是伪造的那个IP了。

**Comment by rprx** [comment]
> 
> 
> 怎么合并到master了

只合并到 v5 的是 breaking，其它的应当先合并到 master，然后同步到 v5

**Comment by rprx** [comment]
@lucifer9 

不确定 nginx 反代 ws 时的行为

**Comment by kslr** [comment]
> > 
> > 
> > 怎么合并到master了
> 
> 只合并到 v5 的是 breaking，其它的应当先合并到 master，然后同步到 v5

不现实，维护不了。

**Comment by rprx** [comment]
> 不现实，维护不了。

这才是最佳的方案吧，breaking 只放 v5，master 正常维护，breaking 攒多了再计划发 v5.0.0。

至少现在 v5 还没有严重分叉，能合并到 master 的却合并到 v5，会打击积极性。

**Comment by kslr** [comment]
> > 不现实，维护不了。
> 
> 这才是最佳的方案吧，breaking 只放 v5，master 正常维护，breaking 攒多了再计划发 v5.0.0。
> 
> 至少现在 v5 还没有严重分叉，能合并到 master 的却合并到 v5，会打击积极性。

这样的话需要不断保持双向合并，是没有必要的。
有关积极性的问题可以通过Pre来改善。

最后为什么V5只放breaking呢？ V5的产生并不是因为要改变一些东西，而是要在时间点之前及时和旧的仓库分割，不论有没有变化，都需要达到这个目标。

**Review by ququ0755 — COMMENTED** [review]
good

**Comment by aybru** [comment]
good

**Comment by rprx** [comment]
@kslr 

1. 首先，即使按照原时间表，现在也应该合并到 master，v5 分支是我昨天提前开的
2. 在此之前，今天已经有两个 PR 合进 master 了，而这个 PR 有特殊的实用性，应该放到 v4
v5 可能会有更大的改变，如随意组合，但正式版得等明年二月，这个 PR 隔这么久是没有必要的
3. 可以发现，breaking 不应阻塞其它功能的正常 release，省事不应该放第一位
4. v5 的定位是人为规定的，即可以改的，不必死板，用 v5 分支来承接 breaking 非常合适

**Comment by kslr** [comment]
如果依旧要保留平时发布流程，不如重新讨论V5计划，而不是确定又改变。

**Comment by rprx** [comment]
~其实昨晚就有说了~

**Comment by lucifer9** [comment]
> @lucifer9
> 
> 不确定 nginx 反代 ws 时的行为

只信任X-Forwarded-For肯定不靠谱的
如果确信客户端是直接连过来的，那X-Forwarded-For只记录对端IP就行，不用非按照标准append，nginx的话是
```
proxy_set_header X-Forwarded-For $remote_addr;
```
如果客户端确实走了代理，比如要过CDN，这样就不能直接记录对端IP，但是这种情况下也不能完全信任X-Forwarded-For毕竟这个可以随便改。一般CDN的公司会有相应的解决方案，比如[cloudflare会用CF-Connecting-IP](https://support.cloudflare.com/hc/en-us/articles/200170986-How-does-Cloudflare-handle-HTTP-Request-headers-)来表示实际发起连接的IP，通常也就是实际客户端IP了。

PROXY protocol的话，cloudflare付费的spectrum是支持的，免费方案不行。反向代理(haproxy)打开accept-proxy以后就没法连接了。

**Comment by lucifer9** [comment]
如果代码里面提高PROXY protocol的优先级呢
就是如果h2/ws启用了PROXY protocol的话
就不再检测header

**Comment by rprx** [comment]
> 
> 
> 如果代码里面提高PROXY protocol的优先级呢
> 就是如果h2/ws启用了PROXY protocol的话
> 就不再检测header

这样处理不太合理，可能会只记录 CDN 的 IP

或许应该把现在的行为改为取最后一个 IP，基本上可以解决客户端伪造 IP 的问题

**Comment by rprx** [comment]
之前提到了不确定 nginx 反代 ws 时的行为，是因为我觉得它好像是直接覆盖，而不是 append

**Comment by lucifer9** [comment]
如果用户已经启用PROXY protocol了，那么以此为优先应该是合理的
毕竟这个是用户主动的行为，相比X-Forwarded-For这种被动行为更能体现用户的意愿

取X-Forwarded-For最后一个IP，如果是通过CDN过来的那么取得的还是CDN的地址

nginx那个，是这样的，因为网上教程大多是写的
```
proxy_set_header X-Forwarded-For $remote_addr;
```
这个当然就reset成最后一个过来的IP了...
而正确（符合标准）的应该是
```
proxy_set_header   X-Forwarded-For    $proxy_add_x_forwarded_for
```
haproxy 默认就是这种行为。
其实header确实很乱的，比如有的代理喜欢再加一个X-Forwarded-For而不是往已有的后面添加，而不是所有的应用都支持处理duplicated的header的，由此又产生奇葩的问题...
也许我们应该在文档中推荐用户在不跨CDN而又有获取真实客户IP的情况下优先使用PROXY protocol
跨CDN的时候尽量参考CDN官方的说明，比如CF的CF-Connecting-IP，也可以在代码中优先支持这种定制化的更准确的header，反正是只能是尽人事听天命了，谁让header可以很容易的伪造呢

**Comment by rprx** [comment]
@lucifer9 

Nginx 那个，刚想把表述改成，“一般会被配置为直接覆盖，而不是 append”

**Comment by rprx** [comment]
@lucifer9 

只取 PROXY protocol 的问题是，比如说我配置了 VLESS 回落到 WS 且开了它，但我套 CDN 时，就只会取到 CDN 的 IP 了

**Comment by rprx** [comment]
@lucifer9 

不管是覆盖还是 append 还是多个 X-Forwarded-For，只取最后一个，就是 Caddy/Nginx/CDN 设置的了，可以解决此问题

**Comment by lucifer9** [comment]
> @lucifer9
> 
> 只取 PROXY protocol 的问题是，比如说我配置了 VLESS 回落到 WS 且开了它，但我套 CDN 时，就只会取到 CDN 的 IP 了

是，这还是刚才说的，需要过CDN的时候，就还是用header比较好，再说CDN也不支持PROXY protocol
所以文档也许应该说明一下PROXY protocol和header各自适用的场合，以及example里面也许可以给出最佳实践



> @lucifer9
> 
> 不管是覆盖还是 append 还是多个 X-Forwarded-For，只取最后一个，就是 Caddy/Nginx/CDN 设置的了，可以解决此问题

这个地方是这样的，如果用户不过CDN，那直接两端启用PROXY protocol就行了，何必管header
如果过CDN，用户还在v2ray前面放了个reverse proxy，那最后一个IP就是reverse proxy设置的，也就是CDN连接过来的IP...

**Comment by rprx** [comment]
@lucifer9 

（没有研究过，能反代 WS 的 Web 软件普遍支持同时 PROXY protocol 吗？）

**Comment by lucifer9** [comment]
> @lucifer9
> 
> （没有研究过，能反代 WS 的 Web 软件普遍支持同时 PROXY protocol 吗？）

apache，nginx，haproxy 都支持
对了，Caddy2 不支持哈哈

**Comment by rprx** [comment]
@lucifer9 

相应地，可以给 `sockopt` 加参数 `"sendProxyProtocol": 0 / 1 / 2`

**Comment by rprx** [comment]
（GitHub 出 BUG 了，上面两条顺序应对调）

**Comment by rprx** [comment]
@lucifer9 

下一步，我们可以允许 VLESS 等协议出站的 `address` 填写 DS 地址，这样就可以在 v5 彻底扔掉 DS 底层传输方式了。

**Comment by rprx** [comment]
https://github.com/v2fly/v2ray-core/pull/367#issuecomment-720054681

看起来 GitHub 的 BUG 仍未结束，这条消息卡 BUG 里了，被不断改时间

**Comment by rprx** [comment]
或许是浏览器导致了这样的 BUG，发上一条消息前刷新了页面，它就不跳了

**Comment by IceCodeNew** [comment]
> "sockopt": {
>       "acceptProxyProtocol": true
> }

看起来这块配置建议还没有被更新到文档中。

**Pull request merged**